wandb/integration/metaflow/metaflow.py

```diff
@@ -0,0 +1,272 @@
+import inspect
+import pickle
+from functools import wraps
+from pathlib import Path
+
+import wandb
+from fastcore.all import typedispatch
+
+from metaflow import current
+
+# I think importing here is more appropriate than importing in the func?
+try:
+    import pandas as pd
+except ImportError:
+    print(
+        "Warning: `pandas` not installed >> @wandb_log(datasets=True) may not auto log your dataset!"
+    )
+
+try:
+    import torch
+    import torch.nn as nn
+except ImportError:
+    print(
+        "Warning: `pytorch` not installed >> @wandb_log(models=True) may not auto log your dataset!"
+    )
+
+try:
+    from sklearn.base import BaseEstimator
+except ImportError:
+    print(
+        "Warning: `sklearn` not installed >> @wandb_log(models=True) may not auto log your dataset!"
+    )
+
+
+class ArtifactProxy:
+    def __init__(self, flow):
+        # do this to avoid recursion problem with __setattr__
+        self.__dict__.update(
+            {
+                "flow": flow,
+                "inputs": {},
+                "outputs": {},
+                "base": set(dir(flow)),
+                "params": {p: getattr(flow, p) for p in current.parameter_names},
+            }
+        )
+
+    def __setattr__(self, key, val):
+        self.outputs[key] = val
+        return setattr(self.flow, key, val)
+
+    def __getattr__(self, key):
+        if key not in self.base and key not in self.outputs:
+            self.inputs[key] = getattr(self.flow, key)
+        return getattr(self.flow, key)
+
+
+@typedispatch  # noqa: F811
+def wandb_track(name: str, data: (dict, list, set, str, int, float, bool), run=None, testing=False, *args, **kwargs):  # type: ignore
+    if testing:
+        return "scalar"
+
+    run.log({name: data})
+
+
+@typedispatch  # noqa: F811
+def wandb_track(
+    name: str, data: Path, datasets=False, run=None, testing=False, *args, **kwargs
+):
+    if testing:
+        return "Path" if datasets else None
+
+    if datasets:
+        artifact = wandb.Artifact(name, type="dataset")
+        if data.is_dir():
+            artifact.add_dir(data)
+        elif data.is_file():
+            artifact.add_file(data)
+        run.log_artifact(artifact)
+        print(f"wandb: logging artifact: {name} ({type(data)})")
+
+
+@typedispatch  # noqa: F811
+def wandb_track(
+    name: str,
+    data: pd.DataFrame,
+    datasets=False,
+    run=None,
+    testing=False,
+    *args,
+    **kwargs,
+):
+    if testing:
+        return "pd.DataFrame" if datasets else None
+
+    if datasets:
+        artifact = wandb.Artifact(name, type="dataset")
+        with artifact.new_file(f"{name}.csv") as f:
+            data.to_csv(f)
```

Comment:
I wonder if we should save the dataframe in a better format that preserved the type information...

Reply:
Now using `pyarrow` to save to parquet